autodiff/reverse/var/var.hpp

```diff
@@ -911,19 +1058,168 @@ struct Hypot3Expr : TernaryExpr<T>
 
     Hypot3Expr(const T& v, const ExprPtr<T>& ll, const ExprPtr<T>& cc, const ExprPtr<T>& rr) : TernaryExpr<T>(v, ll, cc, rr) {}
 
-    virtual void propagate(const T& wprime)
+    void propagate(const T& wprime) override
     {
         l->propagate(wprime * l->val / val);
         c->propagate(wprime * c->val / val);
         r->propagate(wprime * r->val / val);
     }
 
-    virtual void propagatex(const ExprPtr<T>& wprime)
+    void propagatex(const ExprPtr<T>& wprime) override
     {
         l->propagatex(wprime * l / val);
         c->propagatex(wprime * c / val);
         r->propagatex(wprime * r / val);
     }
+
+    void update() override
+    {
+        l->update();
+        c->update();
+        r->update();
+        this->val = hypot(l->val, c->val, r->val);
+    }
+};
+
+// Any expression yielding a boolean depending on arithmetic subexpressions
+struct BooleanExpr
+{
+    template<typename Comparator>
+    struct Negate {
+      template<typename T>
+      bool operator () (const T& a, const T& b) const {
+        return !(Comparator {}(a, b));
+      }
+    };
+
+    bool val = {};
+
+    explicit BooleanExpr(bool v) : val(v) {}
+    operator bool() const { return val; }
+
+    virtual void update() = 0;
+    virtual std::shared_ptr<BooleanExpr> wrap() const = 0;
+};
+
+/// Capture numeric comparison between two expression trees
+template<typename T, typename Comparator>
+struct ExpressionComparison : public BooleanExpr
```

Comment:
Wondering if we could eliminate this type (`ExpressionComparison`) by using a lambda function with mutable capture. 

The lambda function would be created in the body of the comparison operators (`==`, `!=`, `<`, `>=`, etc.). Something like this (not exactly like this!):

~~~c++
template<typename T>
auto operator==(const ExprPtr<T>& l, const ExprPtr<T>& r) 
{
    auto predicate = [=]() mutable 
    {
        l.update();
        r.update();
        return value(l) == value(r);
    };
    
    return BooleanExpr(predicate);
}
~~~

Reply:
Are the semantics of copying a lambda instance as you suggest the same as `shared_ptr`? I'm just wondering about the `propagatex` function of `ConditionalExpr<T>` when (currently) the predicate expression tree gets shared with a new expression containing the derivative of the left and right sides of the expression. If, instead, the lambda instance gets copied into the new gradient expression tree, is that okay?

Also, your suggestion would mix update and evaluate functionality, which could be a performance hit in some use cases. It's separated so far for floating expressions, so why mix for booleans?

Reply:
The lambda functions would be copied along with their captured expression pointers (`ExprPtr`). 

> Also, your suggestion would mix update and evaluate functionality, which could be a performance hit in some use cases. It's separated so far for floating expressions, so why mix for booleans?

Do you mean the `l.update()` and `r.update()` above and then the `value(l)` and `value(r)`? These `value` calls are cheap - just extract the value of the evaluated expressions `l` and `r`. Is there a way to avoid this? My feeling is that these expressions `l` and `r` always need to be evaluated in case there are nodes (independent variables) in these expressions whose values changed.

Reply:
The reason I'm worried is that an `update` call at the root of an expression tree is a non-tail-recursive DFS-like traversal of the entire expression tree with a stack frame per layer without assurances that nodes won't be visited twice if there's structural sharing. I haven't benchmarked anything, but that sounds like a potential problem to me.

Changing the implementation like this by mixing `update` and evaluation doesn't really change much since `update` still has to be called at the root of the expression tree by hand due to the value caching. Changing `value` calls to auto-`update` in all expressions would imply that there's no point to caching anymore.